core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1116,23 +1120,33 @@ class DAGScheduler(
    */
   private[scheduler] def handleTaskCompletion(event: CompletionEvent) {
     val task = event.task
+    val taskId = task.partitionId
     val stageId = task.stageId
     val taskType = Utils.getFormattedClassName(task)
 
     outputCommitCoordinator.taskCompleted(
       stageId,
-      task.partitionId,
+      taskId,
       event.taskInfo.attemptNumber, // this is a task attempt number
       event.reason)
 
-    // The success case is dealt with separately below, since we need to compute accumulator
-    // updates before posting.
-    if (event.reason != Success) {
-      val attemptId = task.stageAttemptId
-      listenerBus.post(SparkListenerTaskEnd(stageId, attemptId, taskType, event.reason,
-        event.taskInfo, event.taskMetrics))
+    // Reconstruct task metrics. Note: this may be null if the task has failed.
```

Comment:
`SparkListenerTaskEnd` should probably mark the `taskMetrics` field as `@Nullable` and also explain that this can happen on failure.